src/marlin_stubs/sdcard/M20-M30_M32-M34.cpp

```diff
@@ -1,22 +1,80 @@
 #include <dirent.h>
 
 #include "../../lib/Marlin/Marlin/src/gcode/gcode.h"
+#include "libs/hex_print_routines.h"
 #include "marlin_server.hpp"
 #include "media.h"
 #include "marlin_vars.hpp"
 
-// M20 - List SD card
-void GcodeSuite::M20() {
-    SERIAL_ECHOLNPGM(MSG_BEGIN_FILE_LIST);
+struct TimeFlags {
+    bool print_lfn : 1;
+    bool print_timestamps : 1;
+};
+
+static void time_to_m20(const struct stat stat_buf, uint16_t *m20_date, uint16_t *m20_time) {
+    time_t t = stat_buf.st_mtim.tv_sec;
+    struct tm lt;
+    localtime_r(&t, &lt);
+    *m20_date = (lt.tm_year + 1900 - 1980) << 9 | (lt.tm_mon + 1) << 5 | lt.tm_mday;
+    *m20_time = lt.tm_hour << 11 | lt.tm_min << 5;
+    *m20_time |= int((lt.tm_sec - (lt.tm_sec % 2)) / 2);
+}
+
+static void print_file_info(const char *const path, const char *const long_path, const struct TimeFlags tf) {
+    // &path[5] and &long_path[5], because we skip the /usb/ prefix on print as it is only used internally
+    SERIAL_ECHO(&path[5]);
```

Comment:
There's a `strlen_constexpr()` in `str_utils.hpp`, we should be able to use that. Not that I am expecting the internal path to change, it just looks more readable/maintainable to me.

Reply:
Indeed, good idea.